core/arch/interp.c

```diff
@@ -5771,6 +5770,170 @@ mangle_x64_ib_in_trace(dcontext_t *dcontext, instrlist_t *trace, instr_t *target
+check_patched_ibl(instrlist_t *trace, instr_t *targeter, int *added_size,
+                  bool *has_str_ldr)
+{
+    int i, j;
+    instr_t *prev = instr_get_prev(targeter);
+    /* XXX: Currently we assume the previous instruction is undecoded. */
+    if (!instr_raw_bits_valid(prev)) {
+        ASSERT_CURIOSITY(false);
+        return DR_REG_NULL;
+    }
+    uint num_instr = prev->length / AARCH64_INSTR_SIZE;
+    if (num_instr < 2)
+        return DR_REG_NULL;
+
+    /* We start from last instruction to scan for str/mov pattern. */
+    for (i = num_instr - 2; i >= 0; i--) {
+        uint prev_prev_enc = ((uint *)prev->bytes)[i];
+        uint prev_enc = ((uint *)prev->bytes)[i + 1];
+        /* We bypass the decoder and compare the raw bytes directly. */
+        if (prev_prev_enc == SAVE_IBL_REG_ENC &&
+            (prev_enc & MOV_IBL_REG_ENC_MASK) == MOV_IBL_REG_ENC) {
+            /* If found, remove these two instructions. */
+            prev->length -= 2 * AARCH64_INSTR_SIZE;
+            /* Concatenate the rest of bytes. */
+            for (j = i; j < num_instr - 2; j++)
+                ((uint *)prev->bytes)[j] = ((uint *)prev->bytes)[j + 2];
+            /* Return jump target register id. */
+            *added_size -= 2 * AARCH64_INSTR_SIZE;
+            return ((prev_enc & MOV_IBL_REG_MASK) >> 16) + DR_REG_START_GPR;
+        }
+        /*
+         * checking for the special case where there isn't a str/mov being patched
+         * rather there is an str/ldr which could happen when for example, the following
+         * basic block is being patched
+         *
+         *  ....
+         *  blr    dr_stolen_reg -> %x30
+         *  b      ibl_routine
+         * into
+         *
+         *  ...
+         *  str tgt_reg, TLS_REG2_SLOT
+         *  ldr tgt_reg, TLS_REG_STOLEN_SLOT
+         *  b ibl_routine
+         *
+         *  This means that we should not remove the str/ldr, rather, we can compare the
+         *  trace_next_exit with tgt_reg directly
+         */
+        else if (prev_prev_enc == SAVE_IBL_REG_ENC && prev_enc == RESTORE_IBL_REG_ENC) {
+            *has_str_ldr = true;
+            return IBL_TARGET_REG;
+        }
+    }
+    return DR_REG_NULL;
+}
+
+/* For AArch64 we have a special case if we cannot remove all code after
+ * the direct branch, which is mangled by cbz/cbnz stolen register.
+ * For example:
+ *    cbz x28, target
+ * would be mangled (see mangle_cbr_stolen_reg() in aarchxx/mangle.c) into:
+ *    str x0, [x28]
+ *    ldr x0, [x28, #32]
+ *    cbnz x0, fall       <- meta instr, not treated as exit cti
+ *    ldr x0, [x28]
+ *    b target            <- delete after
+ * fall:
+ *    ldr x0, [x28]
+ *    b fall_target
+ *    ...
+ * If we delete all code after "b target", then the "fall" path would
+ * be lost. Therefore we need to append the fall path at the end of
+ * the trace as a fake exit stub. Swapping them might be dangerous since
+ * a stub trace may be created on both paths.
```

Comment:
> what would happen if target_cache is pointing to one of these mangled instructions 

The uncompleted mangled app instruction will be re-executed, so all state changes have to be rolled back: restore spilled registers, etc.

Reply:
> > Are you interested in implementing #5062?
> 
> How long do you estimate this would take? I am currently on the final week of an internship where I try to port the trace support for a64. Obviously I can continue contributing after the internship but I am just trying to get a sense of how long this is going to take.

I don't know for sure -- it might be quite straightforward:
+ go enable things permanently that are currently gated on monitor_data_t.pass_to_client like extend_unmangled_ilist's work
+ change internal_extend_trace to estimate the size from the unmangled list (maybe w/ some fudge factor for mangling: for new scheme only used to keep trace below max sizes I would assume)
+ delete monitor_data_t.trace_buf, extend_trace, etc.
+ try to delete decode_fragment: check other callers

But there might be some extra things to deal with once in the details.

Maybe better to finish this PR off as a deliverable -- but the missing translation support is an issue.  Could revert the optimization and have sub-optiimal code that matches x86 with the spill-and-move left there and a simple exit stub and complete translation support, with a comment that we can improve the perf later.

Reply:
> The uncompleted mangled app instruction will be re-executed, so all state changes have to be rolled back: restore spilled registers, etc.

In that case is it OK to include these instructions in the recreated ilist
```
 * trace_exit_label:
 *      ldr x0, TLS_REG0_SLOT
 *      str x2, TLS_REG2_SLOT
 *      mov x2, jump_target
 *      b ibl_routine
 ```
 And update the translation to restore the app state if necessary?

Reply:
> In that case is it OK to include these instructions in the recreated ilist (currently they are not)

Wait -- why are they not in the recreation of the trace?  That is not good.  It should go through the same mangling steps.

Reply:
Because these instructions are added in `fixup_indirect_trace_exit` which is not called when `recreate_fragment_ilist` runs. I have added a call in `recreate_fragment_ilist` so that these instructions are included in the recreated instrlist as well.